sdk/core/azure-core/src/main/java/com/azure/core/http/policy/AzureSasCredentialPolicy.java

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.policy;
+
+import com.azure.core.credential.AzureSasCredential;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Mono;
+
+import java.util.Objects;
+
+/**
+ * Pipeline policy that uses an {@link AzureSasCredential} to set the shared access signature for a request.
+ * <p>
+ * Requests sent with this pipeline policy are required to use {@code HTTPS}. If the request isn't using {@code HTTPS}
+ * an exception will be thrown to prevent leaking the shared access signature.
+ */
+public final class AzureSasCredentialPolicy implements HttpPipelinePolicy {
+    private final AzureSasCredential credential;
+
+    /**
+     * Creates a policy that uses the passed {@link AzureSasCredential} to append sas to query string.
+     *
+     * @param credential The {@link AzureSasCredential} containing the shared access signature to use.
+     * @throws NullPointerException If {@code credential} is {@code null}.
+     */
+    public AzureSasCredentialPolicy(AzureSasCredential credential) {
+        Objects.requireNonNull(credential, "'credential' cannot be null.");
+
+        this.credential = credential;
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        HttpRequest httpRequest = context.getHttpRequest();
+        if ("http".equals(httpRequest.getUrl().getProtocol())) {
+            return Mono.error(new IllegalStateException(
+                "Shared access signature credentials require HTTPS to prevent leaking the shared access signature."));
+        }
+
+        String signature = credential.getSignature();
+        if (signature.startsWith("?")) {
+            signature = signature.substring(1);
+        }
+
+        String query = httpRequest.getUrl().getQuery();
+        String url = httpRequest.getUrl().toString();
+        if (query == null || query.isEmpty()) {
+            url = url + "?" + signature;
+        } else if (!query.contains(signature)) {
```

Comment:
This check is to prevent multiple appends if request is retried. I.e. Storage scopes auth policy per each retry.

Reply:
Based on this should this policy implement https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpPipelinePolicy.java#L36 with `PER_CALL`?

Reply:
I took a look how `getPipelinePosition` is used. It seems to be suited more towards users providing custom policies through client builders. I.e. I don't see any sdk checking this property while adding auth policy to the pipeline. So this would serve as strong enforcement but rather suggestion. Therefore I'd rather keep this check to prevent surprise should some sdk wrongly scope this.
I can override the getPipelinePosition to indicate where it should be placed.

Reply:
For this policy I would override `getPipelinePosition` as we expect to add the SAS token only once for the initial request and any number of retries (unlike shared key where the Date header is taken into account so it needs recalculation per retry).

Reply:
Sounds good. I removed this check and getPipelinePosition is now overriden in this class.